pkgs/build-support/trivial-builders/test-overriding.nix

```diff
@@ -91,7 +91,7 @@ let
   # works at run time.
   runTest = script:
     let
-      name = script.name or (builtins.baseNameOf script);
+      name = script.name or (builtins.unsafeDiscardStringContext (builtins.baseNameOf script));
```

Comment:
Why? (the answer should probably be in a comment)

Reply:
I honestly have no idea. I just asked about the evaluation error in some Matrix channel and was told this incantation.

Reply:
I at least added a comment with the problem description.

Reply:
OK, let me explain how I can, and then we can discuss how to document the issue. 

Nix tries to keep track of dependencies (naturally). In particular, a derivation with builder code referring to some output path of another derivation should have at least a build-time dependency. Moreover, if only the hash+name is taken out of a string to slap together a full path later by concatenating storeDir, Nix wants to track this. So every string comes with provenance information, what derivation output paths have been mangled to produce this string.

For some reason (I am not sure it is an absolutely unavoidable restriction, but the current implementation has it) Nix does _not_ want derivation names to be tainted by the output paths of other derivations.

Here you discard context (i.e. provenance tracking), forgetting which derivation output paths were used to produce the string. Fortunately, it is even «honest» in the sense that you discard the path components where output paths matter anyway.